R/parse_dttm_fmt.R

```diff
@@ -0,0 +1,253 @@
+#'
+#' @returns A regex as a string.
+#'
+#' @keywords internal
+months_abb_regex <- function(x = month.abb, case = c("any", "upper", "lower", "title")) {
+
+  admiraldev::assert_character_vector(x)
+  case <- match.arg(case)
+
+  if (identical(case, "any")) x <- str_to_anycase(x)
+  if (identical(case, "upper")) x <- stringr::str_to_upper(x)
+  if (identical(case, "lower")) x <-  stringr::str_to_lower(x)
+  if (identical(case, "title")) x <- stringr::str_to_title(x)
+
+  stringr::str_flatten(x, collapse = "|")
+}
+
+
+# Date time components. This is a nice
+# utility function that allows you to easily
+# change the regexp for one specific dttm component
+# while keeping the other defaults.
+fmt_c <- function(sec = "S+",
+                  min = "M+",
+                  hour = "H+",
+                  mday = "d+",
+                  mon = "m+",
+                  year = "y+") {
+
+  c(
+    sec = sec,
+    min = min,
+    hour = hour,
+    mday = mday,
+    mon = mon,
+    year = year
+  )
+
+}
+
+regex_or <- function(x, .open = FALSE, .close = FALSE) {
+
+  admiraldev::assert_character_vector(x)
+  admiraldev::assert_logical_scalar(.open)
+  admiraldev::assert_logical_scalar(.close)
+
+  if (.open) x <- c("", x)
+  if (.close) x <- c(x, "")
+
+  stringr::str_flatten(x, collapse = "|")
+}
+
+fmt_rg <- function(
+    sec = "(\\b\\d|\\d{2})(\\.\\d*)?",
+    min = "(\\b\\d|\\d{2})",
+    hour = "\\d?\\d",
+    mday = "\\b\\d|\\d{2}",
+    mon = stringr::str_glue("\\d\\d|{months_abb_regex()}"),
+    year = "(\\d{2})?\\d{2}",
+    na = NULL,
+    sec_na = na,
+    min_na = na,
+    hour_na = na,
+    mday_na = na,
+    mon_na = na,
+    year_na = na) {
+
+  sec_na <-
+    ifelse(!is.null(sec_na), regex_or(sec_na, .open = TRUE), "")
+  min_na <-
+    ifelse(!is.null(min_na), regex_or(min_na, .open = TRUE), "")
+  hour_na <-
+    ifelse(!is.null(hour_na), regex_or(hour_na, .open = TRUE), "")
+  mday_na <-
+    ifelse(!is.null(mday_na), regex_or(mday_na, .open = TRUE), "")
+  mon_na <-
+    ifelse(!is.null(mon_na), regex_or(mon_na, .open = TRUE), "")
+  year_na <-
+    ifelse(!is.null(year_na), regex_or(year_na, .open = TRUE), "")
+
+
+  c(
+    sec = stringr::str_glue("(?<sec>{sec}{sec_na})"),
+    min = stringr::str_glue("(?<min>{min}{min_na})"),
+    hour = stringr::str_glue("(?<hour>{hour}{hour_na})"),
+    mday = stringr::str_glue("(?<mday>{mday}{mday_na})"),
+    mon = stringr::str_glue("(?<mon>{mon}{mon_na})"),
+    year = stringr::str_glue("(?<year>{year}{year_na})")
+  )
+}
+
+# Scalar version of `parse_dttm_fmt()`.
+parse_dttm_fmt_ <- function(x, pattern) {
+
+  match_data <- regexpr(pattern, x)
+  match <- reg_matches(x, match_data)
+
+  is_match <- !is.na(match)
+
+  start <- ifelse(is_match, match_data, NA_integer_)
+  len <- ifelse(is_match, attr(match_data, "match.length"), NA_integer_)
+  end <- start + len - 1L
+  tibble::tibble(pat = pattern, cap = match, start = start, end = end, len = len)
+}
+
+#' @importFrom rlang .data
```

Comment:
Shall we declare all imports in the `package.R`?

Reply:
Perhaps not all but definitely all those imports that are used very often, and I think rlang's `.data` certainly qualifies.

Reply:
Done: https://github.com/pharmaverse/sdtm.oak/commit/ae8966c3800d154c4b4c5c1f9ace7ae47a3a4713.